speakerbox/preprocess.py

```diff
@@ -0,0 +1,580 @@
+    The provided labeled diarized audio directory(s) should have the
+    following structure::
+
+        {labeled_diarized_audio_dir}/
+        ├── label
+        │   ├── 1.wav
+        │   └── 2.wav
+        ├── second_label
+        │   ├── 1.wav
+        │   └── 2.wav
+
+    Generated and attached conversation ids are pulled from the labeled diarized audio
+    directory names.
+    """
+    # Ensure dataset dir is valid
+    audio_output_dir = Path(audio_output_dir).resolve()
+    if audio_output_dir.is_file():
+        raise NotADirectoryError(audio_output_dir)
+    audio_output_dir.mkdir(parents=True, exist_ok=True)
+
+    # Normalize annotation dir to list
+    if isinstance(labeled_diarized_audio_dir, Path):
+        labeled_diarized_audio_dir = [labeled_diarized_audio_dir]
+
+    # Iter over all dia dirs
+    annotated_audios: List[AnnotatedAudio] = []
+    for anno_dir in labeled_diarized_audio_dir:
+        # Convert to path
+        anno_dir = Path(anno_dir).resolve(strict=True)
+
+        # Get the conversation id
+        conversation_id = anno_dir.name
+
+        # Iter over all labels
+        for label_dir in anno_dir.iterdir():
+            if label_dir.is_dir():
+                for i, audio_file in enumerate(label_dir.glob("*.wav")):
+                    audio = AudioSegment.from_file(audio_file)
+                    monologue_speaker = label_dir.name
+
+                    # Construct basic filename
+                    filename_base = (
+                        f"{audio_file.with_suffix('').name}"
+                        f"-monologue_{i}-{monologue_speaker}"
+                    )
+
+                    # Chunk audios for this monologue
+                    annotated_audios = _chunk_audio(
+                        monologue_start_time=0,
+                        monologue_end_time=audio.duration_seconds,
+                        max_audio_chunk_duration=max_audio_chunk_duration,
+                        min_audio_chunk_duration=min_audio_chunk_duration,
+                        audio=audio,
+                        audio_output_dir=audio_output_dir,
+                        chunk_filename_base=filename_base,
+                        overwrite=overwrite,
+                        annotated_audios=annotated_audios,
+                        conversation_id=conversation_id,
+                        label=monologue_speaker.lower(),
+                    )
+
+                log.debug(f"Completed expansion for diarized audio dir: {label_dir}")
+
+    # Merge all into a single dataframe
+    return pd.DataFrame([aa.to_dict() for aa in annotated_audios])  # type: ignore
+
+
+###############################################################################
+# Data expections and cleaning
+
+
+def prepare_dataset(
+    dataset: pd.DataFrame,
+    test_and_valid_size: float = 0.4,
+    equalize_data_within_splits: bool = False,
+    n_iterations: int = 100,
+    seed: Optional[int] = None,
+) -> Tuple[DatasetDict, pd.DataFrame]:
```

Comment:
> * roughly uniform distribution of labels (in each of the audio files, and thus also in each of the dataset splits - train, test, valid)

From my large model tests this doesn't seem to have much of an effect / we may want to purposfully bias the dataset. I.e. if Gonzalez talks a majority of the time because she is council president, the model should be confident that she is talking a majority of the time too. But this is exactly why this parameter was added. I can attempt to train a large model with this parameter turned on.

> Seems like `equalize_data_within_splits` is going to throw out some data points?

Yes it does so at random, per split.

> Maybe we could ask user (in the README) to use audio files where each speaker has roughly the same amount of speaking time or as close as possible?

Yep, I can at least add that documentation, the best example of this for us was using the council briefing meetings which are the "what am I doing this week in council" where everyone goes around and shares. The sitting council president naturally has higher speaking times for any meeting, including briefings, however.

Reply:
> we may want to purposfully bias the dataset. I.e. if Gonzalez talks a majority of the time because she is council president, the model should be confident that she is talking a majority of the time too. But this is exactly why this parameter was added. I can attempt to train a large model with this parameter turned on.

I don't think that's needed. If we don't want the uniform distribution of labels, then I think the same distribution across train, test, and valid is good enough.